teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1071 @@
+```
+</td>
+</tr>
+</table>
+
+Referencing `StepAction` in a `Task` could create name conflicts that need to be resolved by the `Task` author. For example, if “step1” needs a `Parameter` called “url” and “step2” also needs a `Parameter` called “url” but they are actually meant for different purposes, the `Task` author needs to be able to pass in different values to both these `Parameters`.
+
+
+All inlined `Steps` and `Sidecars` in a `Task` can access all the `Parameters` and reference them in their work. This behavior is still intact. However, `Parameters` need to be passed explicitly to a referenced `Step`. While users only need to provide values when there are name conflicts (since a `Step` has access to all params of a `Task`) we start by explicitly providing all the required parameters to the `StepAction` and will be enforced during validation. To prevent increased verbosity of the specifications, we can invoke parameter propagation in future work. 
+
+<table>
+<tr>
+<td>
+
+```yaml
+# This is how a step author would
+# declare the params that it needs.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: url
+      description: “This is a param”
+      default: bar
+  image: alpine
+  script: |
+    echo $(params.url)
+
+```
+</td>
+<td rowspan=2>
+
+```yaml
+# This is how a task author would provide 
+# the params that the step needs.
+
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  params:
+    - name: url1
+    - name: param2
+  steps:
+    - name: step1
+      stepRef: 
+        name: myStep
+      params:
+        - name: url
+          value: url1
+    - name: step2
+      stepRef:
+        name: myOtherStep
+      params:
+        - name: url
+          value: url2
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: url
+      description: “Another param”
+      default: baz
+  image: alpine
+  script: |
+    echo $(params.url)
+
+```
+</td>
+</tr>
+</table>
+
+#### Results
+
+A referenced `StepAction` needs to declare if it produces a `Result` so that `Task` authors know that they can reference it without having to drill into the `StepAction` specification. 
+
+If multiple `StepActions` produce the different `Results` under the same name then to ensure that they are not overwritten, the `Task` needs to be able to capture them under different names. In order to capture these at the task level, we need to: 
+1. Introduce a field to [TaskResults](https://github.com/tektoncd/pipeline/blob/445734d92807a80158b4b7af605d768c647fdb3d/pkg/apis/pipeline/v1/result_types.go#L19-L49) called `Value` like we do for [PipelineResults](https://github.com/tektoncd/pipeline/blob/445734d92807a80158b4b7af605d768c647fdb3d/pkg/apis/pipeline/v1/pipeline_types.go#L129C1-L144C2). (Note this would not be a backwards incompatible change. We would be making an additive change here.)
```

Comment:
Completely agreed! I think we can visit this as future work (i.e. `promoted results`). 

For now, a good compromise might be to make the step results available to Task results by default with the expectation that if there are name conflicts then users have to explicitly request them using the above suggested syntax. If users don't resolve the names then they will be overwritten by the following step. 

So if an author uses two StepActions in a task that produce a result called "digest" and a third StepAction that produces a result called "foo" then "foo" would be propagated to the Task by default and the Task author would only need to explicitly call out the result "digest" using the above syntax. 

This would be similar to params where instead of explicitly declaring all params, they only provide params that have name conflicts.

Reply:
Sounds good